### PR TITLE
Hidden field behavior

### LINKED ISSUE #93: Hidden fields get saved
When a field is hidden, the data is still saved off in firebase

### PR DESCRIPTION
**Pull request by ctcleary**
As requested in #93 the behavior is now:
* A user enters data into a field "X" e.g. "This is an answer"
* "X" becomes hidden.
* `Save` action does not persist the data in "X"
* "X" becomes visible, "This is an answer" is still populated in the now-shown field.

To do this I had to entirely rework the hiding mechanism. The per-type hiding behavior is no longer needed. Since every input is constructed through `/types/children.js`, the hiding behavior is now mostly centralized there.
* The field's `field.show` object is assigned a `field.show.hidden` boolean so we can determine when its hidden status has changed.
* Whenever a field's hidden status changes, we register it with the controller from `content-edit/index.js` since this component's data is what gets sent to the `save` action.
* In the `save` action, we create a copy of the current data and set any hidden field's data to `null` in the clone, this filtered data is what ends up persisted to the datastore.

[Fixes #93]

### CODE REVIEW

**Comment by ctcleary** [comment]
Doh. Test failures and travis is yelling about style stuff.

File: src/types/children.js [hunk 1]
```diff
@@ -3,33 +3,55 @@ import get from "lodash.get";
...
+        }
 
-            return m.component(component, assign({}, options, {
-                field : field,
-                class : css[index ? "field" : "first"],
-                data  : get(options.data, field.key),
-                path  : options.path.concat(field.key)
-            }));
-        })
+        if(!component) {
```
Comment (tivac) [inline src/types/children.js]: This check seems like it should go above any field visibility bits. No reason to even bother if it's an unrecognized component type.

**Comment by tivac** [comment]
Why are the hidden fields tracked/modified in `content-edit/index.js`? I don't see any place where it ever actually uses that data.

File: README.md [hunk 2]
```diff
@@ -125,6 +125,34 @@ You can attach dom-specific attributes to most types by specifying an `attrs` ke
 }
 ```
 
+### Conditionally Hidden / Shown Fields
```
Comment (tivac) [inline README.md]: This documentation has moved to `docs/schema.md`.

**Comment by ctcleary** [comment]
The hidden fields are tracked in `content-edit/index.js` because that's where all other field data is tracked. Then during the `content-edit/head.js : ctrl.save()` operation, we call `filterHidden()` using the data tracked via `content-edit/index.js : ctrl.registerHidden()` to remove any currently-hidden field's data from the update. This is the only way (afai can tell) we can retain any information in a hidden field for the desired UX (hiding/unhiding a field does not clear it) without accidentally saving the data from a hidden field when the save button is hit.

File: src/types/children.js (hunk 1 above)
Comment (ctcleary) [inline src/types/children.js]: Yup. Totally right. Super obvious, just missed it.

**Comment by tivac** [comment]
Ah, sure. I need to stew on they way that's wired up for a bit because right now it's very bodged together and I forgot just how clunky it was.

I'd like the tests to be cleaned up (either removed if now-irrelevant or fixed) and then I think we're good to merge. This will need to be rebased on master before that can happen, but that should be pretty small.

**Comment by tivac** [comment]
PS I said this in-person but for posterity: per-type hiding logic was really stupid and I'm glad it's removed.

**Comment by ctcleary** [comment]
I would also like to clean it up a bit, but for now I was shooting to get the new hiding behavior wired up without a total overhaul of any other important systems. Was waiting to update the tests until I got the :+1:  so I'll get on that part probably tomorrow AM.

**Comment by ctcleary** [comment]
This was repeatedly delayed because every time I tried to figure out why I couldn't write tests for `types/children.js` I would run into really bizarre errors I couldn't figure out and eventually had to put it aside again before I lost my mind and bloody rage vomited at my computer like a 28 Days Later zombie.

Turns out the problem is, since `types/children.js` imports half the project, it ends up importing some that assume the presence of `window`, `document`, and `require`. Since these do not (as far as I can tell( exist in Mocha's headless browser, Mcha just pukes and refuses to work when it hits them.

I tried to go throw some "just for testing" patches over these spot, but I think I've determined that the problem is pervasive enough that this is a poor course of action.

I'm inclined to say we skip tests on `types/children.js` for now so that we can get this preferred functionality in. I'm inclined to say we punt testing forward and make a task for making `types/children.js` more testable by reducing its imports, but I also know that's not super likely to happen.

So, ball's in your court, @tivac -- What'chu'wan'me'do?

**Comment by tivac** [comment]
¯\\\_(ツ)\_/¯

Skipping tests on `children.js` is fine for now.

**Comment by ctcleary** [comment]
Tests removed. I think most of the style warnings precede me?

Ready for merge?

File: src/types/children.js [hunk 3]
```diff
@@ -3,33 +3,55 @@ import get from "lodash.get";
...
-            return m.component(component, assign({}, options, {
-                field : field,
-                class : css[index ? "field" : "first"],
-                data  : get(options.data, field.key),
-                path  : options.path.concat(field.key)
-            }));
-        })
+        result = m.component(component, assign({}, options, {
+            field : field,
+            class : addClasses(field, css[index ? "field" : "first"] ),
```
Comment (tivac) [inline src/types/children.js]: Use CSS composition instead of array.join(" ") whenever possible.

**Pull request merged**